### PR TITLE
Add support for ObjectDetection

### LINKED ISSUE #442: [Feature Request] Object Detection Trainer
Hi, I'm curious if there are any plans to add an object detection trainer?

Also, I wonder how feasible it may be to extend this library to work with video data and multi-object tracking tasks? Or is torchvideo more appropriate for this task?

### PR DESCRIPTION
**Pull request by ashnair1**
Adds a new `ObjectDetectionTask` which uses torchvision's `Faster-RCNN` under the hood.

Closes #442

### CODE REVIEW

**Comment by ashnair1** [comment]
> For torchvision dependency version, see what I did in https://github.com/microsoft/torchgeo/pull/631 for the other trainers. It's possible to check the version of torchvision and use different syntax for each version. It's kinda ugly and I don't like it, but I'm not sure if it's wise to just drop support for all older torchvision versions. Open to feedback on this though.

Saw this snippet. 
```
if parse(torchvision.__version__) >= parse("0.12"):
    if backbone_pretrained:
        kwargs = {
            "weights": getattr(
                torchvision.models, f"ResNet{backbone[6:]}_Weights"
            ).DEFAULT
        }
    else:
        kwargs = {"weights": None}
  else:
    kwargs = {"pretrained": backbone_pretrained}
```

This can work. Since the number of backbones supported also include the `ResNext` and `Wide_Resnet` variants, I opted to create  a mapping between backbones and weights. Perhaps this could be moved to a common location in a later PR.


Agree with not dropping support. We only need to drop a version if it breaks something or if it requires really ugly code to handle it. Think we're good for now. 

> For the torchmetrics dependency version, what was the reason for the version bump? It seems like MeanAveragePrecision was introduced in 0.7.0 (renamed from MAP) according to the changelog.

You're right. Think it was just the filename that was changed (`map.py` -> `mean_ap.py`). Reverted that commit.

**Comment by adamjstewart** [comment]
Another option would be to change the `pretrained` kwarg on our end so that instead of taking a boolean it takes whatever torchvision wants. Then it's up to the user to decide which pretrained model to use. Thoughts?

**Comment by ashnair1** [comment]
> Another option would be to change the pretrained kwarg on our end so that instead of taking a boolean it takes whatever torchvision wants. Then it's up to the user to decide which pretrained model to use. Thoughts?

Sure, that could work. That way the user would be able to specify the weights to be used and if they leave it unspecified, we can just use the default weights.

**Comment by adamjstewart** [comment]
Hmm, but then it becomes difficult to store this configuration in a YAML file...

**Comment by ashnair1** [comment]
Right. The YAML would only be informative if you're using torchvision > 0.12

**Comment by adamjstewart** [comment]
But even if the YAML is torchvision > 0.12 specific, can you store objects in YAML or only strings/floats/bools? We could store a string and exec it but that feels like a security risk...

**Comment by ashnair1** [comment]
The weights in the YAML can be string. We could just specify it like this.

``` diff
experiment:
  task: "nasa_marine_debris"
  name: "nasa_marine_debris_test"
  module:
    detection_model: "faster-rcnn"
    backbone: "resnet50"
+   pretrained: "ResNet50_Weights.IMAGENET1K_V1"
    num_classes: 2
    learning_rate: 1.2e-4
    learning_rate_schedule_patience: 6
    verbose: false
  datamodule:
    root_dir: "data/nasamr/nasa_marine_debris"
    batch_size: 4
    num_workers: 56
    val_split_pct: 0.2
```

The pretrained arg can then be passed into the `weights` arg

**Comment by adamjstewart** [comment]
Ah, we could use `torchvision.prototype.models.get_weight` to convert the string to the enum. But yeah, it would be a bit frustrating for the YAML to only work properly with torchvision 0.12.

Let's keep it like this for now but strongly consider dropping support for older torchvision at some point.

**Comment by ashnair1** [comment]
The `pretrained` arg will officially be dropped in version `0.15`. Perhaps that would be a good time to drop old versions and peg minimum to `0.13`.

**Comment by ashnair1** [comment]
Coverage drop is mainly caused by the code required for supporting minimum versions. 

On that note, has anybody looked into combining the coverage reports for normal tests and minimal tests via [`coverage combine`](https://coverage.readthedocs.io/en/6.4.3/cmd.html#combining-data-files-coverage-combine) before uploading to codecov?

**Comment by adamjstewart** [comment]
We run tests on both the latest version of dependencies and the minimum version of dependencies. Both tests generate coverage reports, and should automatically be merged when uploaded to codecov. This works on all of our other PRs, not sure why coverage is dropping for you.

**Comment by ashnair1** [comment]
This is weird. The coverage drop seems to be caused by unrelated files. Also a warning is shown stating there are unmerged base commits. Any ideas?

image

**Comment by adamjstewart** [comment]
> The coverage drop seems to be caused by unrelated files.

This usually means that coverage has increased in main since this branch was created. Usually, rebase or merge commits solve this. Wish codecov was smarter about this...

**Comment by ashnair1** [comment]
Since I've opened this PR, there have been a total of 4 commits merged to main. The two tutorial commits, one related to Sentinel2 and the torchvision constraint fix so it's unlikely a coverage increase is the reason.

I wonder if this has to do with me rebasing and squashing commits in this branch.

**Comment by adamjstewart** [comment]
Yep, I have no idea what's going on here. It's just this PR, right? Probably best to reach out to codecov and see if this is a bug in codecov.

I'm focusing on 0.3.1 stuff at the moment, but if this is still broken after 0.3.1 is finished I can try to report it to codecov myself.

**Comment by isaaccorley** [comment]
I went down the torchvision.detection rabbit hole awhile back. I think my only complaint with it was that they don't support negative sample images that don't have any objects in them during training. Do you know if this has changed? I need to catch up on all the releases.

**Comment by isaaccorley** [comment]
Also I believe they added support for constructing a resnet fpn backbone from a pretrained torchvision resnet model. @calebrob6 and @anthonymlortiz, this might be of interest because we could add support for pretrained SSL resnet backbones to this task.

**Comment by ashnair1** [comment]
@isaaccorley

> I went down the torchvision.detection rabbit hole awhile back. I think my only complaint with it was that they don't support negative sample images that don't have any objects in them during training. Do you know if this has changed? I need to catch up on all the releases.

They support negative samples. Support was added in [`0.6.0`](https://github.com/pytorch/vision/releases/tag/v0.6.0) via pytorch/vision@e75b497


> Also I believe they added support for constructing a resnet fpn backbone from a pretrained torchvision resnet model. @calebrob6 and @anthonymlortiz, this might be of interest because we could add support for pretrained SSL resnet backbones to this task.

This is supported as well. Just need to pass the backbone name and the `pretrained` flag. 

They recently added (in `v0.13.0`) the Multi Weights API which allows you to select which weights your backbone should load. Adam opened an issue to track this - #762 


I really wish `torchvision` would add support for oriented/rotated bounding boxes i.e. `(x, y, w, h, a)`. Orientation is a significant factor when you're dealing with object detection in RS imagery. Especially if your objects are small and clustered together. `detectron2` supports [RotatedBoxes](https://detectron2.readthedocs.io/en/latest/_modules/detectron2/structures/rotated_boxes.html) and the necessary ops like [rotated nms](https://detectron2.readthedocs.io/en/latest/_modules/detectron2/layers/nms.html), [rotated ROIAlign](https://detectron2.readthedocs.io/en/latest/_modules/detectron2/layers/roi_align_rotated.html), [Rotated RPN](https://github.com/facebookresearch/detectron2/blob/main/detectron2/modeling/proposal_generator/rrpn.py) etc and it works really well. Was able to get some nice results on the DOTA dataset using it. 

There is an open issue regarding porting over the necessary ops to `torchvision` (pytorch/vision#2761) but there doesn't seem to be much interest :disappointed:.

**Comment by isaaccorley** [comment]
> I really wish `torchvision` would add support for oriented/rotated bounding boxes i.e. `(x, y, w, h, a)`. Orientation is a significant factor when you're dealing with object detection in RS imagery. Especially if your objects are small and clustered together. `detectron2` supports [RotatedBoxes](https://detectron2.readthedocs.io/en/latest/_modules/detectron2/structures/rotated_boxes.html) and the necessary ops like [rotated nms](https://detectron2.readthedocs.io/en/latest/_modules/detectron2/layers/nms.html), [rotated ROIAlign](https://detectron2.readthedocs.io/en/latest/_modules/detectron2/layers/roi_align_rotated.html), [Rotated RPN](https://github.com/facebookresearch/detectron2/blob/main/detectron2/modeling/proposal_generator/rrpn.py) etc and it works really well. Was able to get some nice results on the DOTA dataset using it.
> 
> There is an open issue regarding porting over the necessary ops to `torchvision` ([pytorch/vision#2761](https://github.com/pytorch/vision/issues/2761)) but there doesn't seem to be much interest 😞.

@conor-wallace was also able to get good benchmark results on DOTA with detectron2 and we use it a ton for aerial imagery now. Though we would love to never touch detectron2 again if torchvision added rotated oriented support.

File: evaluate.py
Comment (adamjstewart) [inline evaluate.py:119]: I wonder if `torchgeo.datasets.unbind_samples` could help here

File: evaluate.py
Comment (adamjstewart) [inline evaluate.py:205]: Can you get rid of this empty line so it's clear that these are all clauses of the same if-statement?

File: evaluate.py
Comment (adamjstewart) [inline evaluate.py:270]: Same here

File: torchgeo/datamodules/nasa_marine_debris.py
Comment (adamjstewart) [inline torchgeo/datamodules/nasa_marine_debris.py:146]: Any chance we can simplify this? Maybe `s/dataset/self.dataset` right before we perform the split, so we can always access `self.dataset` at any time?

File: torchgeo/datasets/nasa_marine_debris.py
Comment (adamjstewart) [inline torchgeo/datasets/nasa_marine_debris.py:106]: How many of the boxes were invalid?

@isaaccorley recently had to do the same thing for IDTReeS, not sure if we should try to reuse code for that.

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:24]: What is the type ignore here for?

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:26]:
```suggestion
if parse(torchvision.__version__) >= parse("0.13"):
```

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:202]:
```suggestion
        """Compute test MAP.
```
?

**Review by adamjstewart — COMMENTED** [review]
This looks great, just a few minor comments to see if we can simplify things a bit.

You can ignore the coverage decrease for now. Before merging I'll run the tests locally just to make sure that they're all false negatives and force merge if I have to. Still quite curious why it isn't working though.

File: torchgeo/trainers/detection.py
Comment (ashnair1) [reply torchgeo/trainers/detection.py:24]: The following mypy error gets triggered without it 

```
torchgeo/trainers/detection.py:24: error: Name "MeanAveragePrecision" already defined (possibly by an import)  [no-redef]
```

File: evaluate.py
Comment (ashnair1) [reply evaluate.py:119]: Think it could. I'll take a look.

Comment (ashnair1) [reply evaluate.py:119]: It can work but we will still need to iterate over them and move them to specified device.

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [reply torchgeo/trainers/detection.py:24]: Can we use `# type: ignore[no-redef]`? I try to be as specific as possible with type ignores. Doesn't matter if it causes the line to take up multiple lines.

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:51]: Needs `versionadded`

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:62]:
```suggestion
                if parse(torchvision.__version__) >= parse("0.13"):
```

**Review by adamjstewart — DISMISSED** [review]
LGTM, will merge after #780 is merged

File: torchgeo/datasets/nasa_marine_debris.py
Comment (isaaccorley) [reply torchgeo/datasets/nasa_marine_debris.py:106]: We can reuse the IDTReeS._filter_boxes method and make it a function so it can be used for both datasets. Doesn't have to be in this PR but it seems as though this may be a common issue with object detection datasets.

**Comment by adamjstewart** [comment]
Can you rebase to resolve the merge conflict?

Still no idea why codecov has been buggy lately. I'll force merge if I need to.

**Comment by adamjstewart** [comment]
The drop in coverage here is legit, I was able to reproduce it locally:
```console
$ pytest --cov=torchgeo/trainers --cov-report=term-missing tests/trainers/
=============================================================================================== test session starts ================================================================================================
platform darwin -- Python 3.9.13, pytest-7.1.3, pluggy-1.0.0
rootdir: /Users/ajstewart/torchgeo, configfile: pyproject.toml
plugins: anyio-3.6.1, nbmake-1.3.3, cov-3.0.0
collected 57 items                                                                                                                                                                                                 

tests/trainers/test_byol.py ....                                                                                                                                                                             [  7%]
tests/trainers/test_classification.py ..................                                                                                                                                                     [ 38%]
tests/trainers/test_detection.py ...                                                                                                                                                                         [ 43%]
tests/trainers/test_regression.py ....                                                                                                                                                                       [ 50%]
tests/trainers/test_segmentation.py ..................                                                                                                                                                       [ 82%]
tests/trainers/test_utils.py ..........                                                                                                                                                                      [100%]

---------- coverage: platform darwin, python 3.9.13-final-0 ----------
Name                                  Stmts   Miss  Cover   Missing
-------------------------------------------------------------------
torchgeo/trainers/__init__.py             9      0   100%
torchgeo/trainers/byol.py               141      1    99%   326
torchgeo/trainers/classification.py     164      0   100%
torchgeo/trainers/detection.py          101      5    95%   24, 66-68, 187-188
torchgeo/trainers/regression.py          87      1    99%   45
torchgeo/trainers/segmentation.py        99      0   100%
torchgeo/trainers/utils.py               52      0   100%
-------------------------------------------------------------------
TOTAL                                   653      7    99%


=============================================================================================== 57 passed in 52.34s ================================================================================================
```
See https://app.codecov.io/gh/microsoft/torchgeo/pull/758 for the pictorial representation.

**Comment by calebrob6** [comment]
image

the try/except in the validation_step

**Comment by adamjstewart** [comment]
There's another line too.

**Comment by ashnair1** [comment]
The try except block in the val_step is the only one line left. Not sure how to test this. I'm guessing `test_segmentation.py` passes because it includes datamodules that don't have `plot` implemented so it would trigger the exception.

**Comment by isaaccorley** [comment]
Maybe instead of try/except we should check if the datamodule has a plot method. Something like: if isinstance(getattr(datamodule, "plot"), Callable). Then in the test we can override the plot method to be None or something so it will cover both cases.

**Comment by adamjstewart** [comment]
There's a reason for the try-except. Lightning is very much dynamically typed, and makes no guarantees that any of these attributes will exist:

* `self.trainer.datamodule`
* `datamodule.plot`
* `self.logger.experiment`

That's also the reason that we have to ignore mypy warnings on so many lines. The correct solution would be to use hasattr/getattr on all of these like @isaaccorley suggested. I was just lazy and wrapped everything in a single try-except. 

If you want to rework the code to use hasattr/getattr, I would create a new PR and do the same for existing trainers. We should probably add plot methods to all datamodules like we have to all datasets, so that will cause this same issue in other tests too.

If you want to keep using the try-except, my suggestion would be to add a test that monkeypatches the datamodule to remove the plot method. This will give you full coverage and ensure that the trainer works even for datamodules without a plot method.

File: tests/trainers/test_detection.py
Comment (adamjstewart) [inline tests/trainers/test_detection.py:40]: Let's put this in a separate test function, there's no reason we need to test this for every datamodule.

Comment (ashnair1) [reply tests/trainers/test_detection.py:40]: How do you mean? Add this separately as `test_missing_attribute` because we don't want to mix this with `test_trainer`?

Currently `test_detection.py` only tests the `NasaMarineDebrisDatamodule`.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
@isaaccorley does this need a predict step?

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:150]: According to https://github.com/microsoft/torchgeo/pull/790#discussion_r984943775 we should use `self()` instead of `self.forward()`. I'll submit a PR to change existing trainers, but can you make that change here?

**Review by adamjstewart — DISMISSED** [review]

**Comment by isaaccorley** [comment]
> @isaaccorley does this need a predict step?

Ideally yes, we can have a predict step that returns the predicted boxes and class labels per sample.

**Comment by adamjstewart** [comment]
Let's add that here before we forget, I want to make sure all of our trainers support the same things.

**Comment by isaaccorley** [comment]
@ashnair1 Let me know if you want some help on it. I can walk you through the predict_steps I made for the classifiers.

**Comment by ashnair1** [comment]
Added `predict_step` and corresponding test.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
